### PR TITLE
Make all mpl styles json loads and update visualization code and docs

### LINKED ISSUE #5169: Make mpl colors and other styles fully configurable by users with a json load
### What is the expected enhancement?

In #5063, the IQX color scheme was added as an option to the mpl drawer. The style dict entry, 'displaycolor' was expanded to include new gates and to allow for entering both gate and text color for each gate type. This expands color choices to include dark gate colors with light text. This first phase was done by creating 3 classes in qcstyle.py.

In this next phase, this can be reduced to 1 default class and the additional styles would be read in from json files. I don't believe users should be allowed to edit these 3 json files - default, iqx, and bw - directly, but they can copy these and create new styles of their own.

This issue proposes to make default, iqx, and bw json style files part of qiskit, perhaps in a styles directory under the visualization directory. User json files should go in a fixed place, perhaps in a styles directory under ~/.qiskit.

The code for reading in the json files already exists, though there is only minimal error checking currently. Deciding on the locations for the files is the main issue.

We might want to look at any possible special cases, like the controlled classical gates in iqx and see if any of these can be addressed in a generic way. Right now all a user can do with color is change the colors of the specific gate names in 'displaycolor'.

Any of the other documented style dict items can be set in the user json file. The user can create a partial style dict in the file, and only those elements entered will be changed for that style.

The user interface is simple - `qc.draw('mpl', style='user_style1.json')`

### PR DESCRIPTION
**Pull request by enavarro51**
### Summary

Fixes #5169

### Details and comments

This PR fully implements style dictionaries in the 'mpl' drawer as JSON files. qcstyle.py now contains a single DefaultStyle class. If the user selects the 'iqx' or 'bw' color scheme, it is read in from a JSON file. The 'default' style is now a dictionary which simplifies the import of other styles. Users can create their own JSON files which can be selected using the 'style' param.

In addition to the functional changes, extensive documentation changes were made to 'circuit_drawer' and 'QuantumCircuit.draw' which included cleaning up language and bringing all the parameters and style dict items up-to-date with the current state of the code.

Some structural changes were made to 'matplotlib.py', which included variable and method name changes and some reordering of the code, especially in the `__init__`. These changes were made to simplify future maintenance of the code.

These are the changes,

- First and foremost is that all of these changes should be 100% backward compatible. No functionality has been removed and no changes should cause a user to have to alter previous calls to any of the drawers.
- The docs in 'circuit_drawer', which are mostly duplicated in 'QuantumCircuit.draw', were out-of-date with the state of the code. These are now current.
- There were 3 items in the 'style' dict that were not implemented in any of the code - 'latexdrawerstyle', 'compress', and 'usepiformat'. Discussions indicate these are not needed, so they have been removed. If a user had one of these in a 'style' dict, they will receive a warning saying it's not valid. Since there was no function for the item, the function of the drawer will not change.
- Some parameters were being passed to the drawers, but never used. An example is 'reverse_bits'. 'circuit_drawer' uses this in the call to '_get_layered_instructions', which is the only place it was needed. But 'reverse_bits' was then being passed to the 'mpl' and 'latex' drawers where it was ignored. Now only relevant parameters are passed from 'circuit_drawer' down through to the final drawer classes.
- There was apparently an intention when the 'latex' drawer was split off from the main drawer, to use the 'BWStyle' from 'qcstyle.py'. 'BWStyle' was loaded at the top of the 'latex' drawer, but never used. The style load has been removed.

- There is now 1 class in 'qcstyle.py' which is 'DefaultStyle'. This class contains 2 dicts, 'colors' and 'style'. 'colors' is only there for informational purposes, as a way to indicate which is the 'pauli' color, for example. The 'style' dict contains all the 'style' items for the 'mpl' drawer.
- The 'default' style is always instantiated first and then if another style has been selected, it will be read in from a JSON file and will overwrite 'default' style items with any items from the JSON file. This is how the 'bw' and the new 'iqx' styles are loaded and the same method will work for any user styles.
- The path search for styles works as follows,
   - There is a new directory, '<qiskit location>/qiskit-terra/qiskit/visualization/styles' which contains the 'iqx' and 'bw' JSON files. This is searched first.
   - There is a new config file item in 'settings.conf' called 'circuit_mpl_style_path'. This can be set to one or more directories to search separated by ';'. For example, 'circuit_mpl_style-path = /home/user;/home/user/styles'. These directories are searched next in order.
   - Finally if none was found previously, the search is done for the raw JSON file name, e.g. 'user_style1.json'.
- Full error trapping now occurs on JSON file loads. For all errors, a warning message is displayed indicating the error and then informing the user that the 'default' style will be used. This means no JSON errors cause failure of the drawer.
- In addition to the obvious 'FileNotFound', 'JSONDecodeError' is also trapped and the warning message will indicate the line and column number of the faulty JSON.
- The previous functionality of passing a dict to the 'style' parameter in 'circuit_drawer' or 'draw' works as it did before, and these style changes are done after 'default' is loaded and after any JSON file loads.
- 3 new 'mpl' tests with images were added - 'reverse_bits', 'bw', and 'user_style', along with tests for the new user config items.

Here is a sample output using a JSON loaded user style. In addition to the obvious color changes, index numbers are added, the right margin has been increased, and the 'Phase' name was changed using 'displaytext'. LaTeX could be used for these names.

image

### CODE REVIEW

File: test/python/test_user_config.py [hunk 1]
```diff
@@ -131,6 +129,7 @@ def test_all_options_valid(self):
         [default]
         circuit_drawer = latex
         circuit_mpl_style = default
+        circuit_mpl_style_path = ~;~/.qiskit
```
Comment (1ucian0) [inline test/python/test_user_config.py:132]: Does `~` means `$HOME`? Does this works in Windows?

Comment (enavarro51) [reply test/python/test_user_config.py:132]: First off, it passes all the auto tests. I'm using this (from the os.path docs)
`os.path.expanduser(path)`
`On Unix and Windows, return the argument with an initial component of ~ or ~user replaced by that user’s home directory.`
It does leave the question of the Mac, but it does pass the Mac tests. So not sure if it's universal.
It's just a testing issue. A user would put in the appropriate directory format for their system. Is there a more generic way to put in 2 dir names?

Comment (1ucian0) [reply test/python/test_user_config.py:132]: I think it's safe. We used the same approach in other parts. 

> Is there a more generic way to put in 2 dir names?

I would use `:`. It's the `$PATH` way.

Comment (enavarro51) [reply test/python/test_user_config.py:132]: Ah, good catch. I'll change that.

File: qiskit/user_config.py [hunk 2]
```diff
@@ -29,14 +30,17 @@ class UserConfig:
     [default]
     circuit_drawer = mpl
     circuit_mpl_style = default
+    circuit_mpl_style_path = /home/user/styles;/home/user
```
Comment (1ucian0) [inline qiskit/user_config.py:33]: I would go like something like this here, right?

```suggestion
    circuit_mpl_style_path = ~/.qiskit:<default location>
```

File: qiskit/circuit/quantumcircuit.py [hunk 3]
```diff
@@ -1130,173 +1130,223 @@ def draw(self, output=None, scale=None, filename=None, style=None,
              interactive=False, plot_barriers=True,
              reverse_bits=False, justify=None, vertical_compression='medium', idle_wires=True,
              with_layout=True, fold=None, ax=None, initial_state=False, cregbundle=True):
-        """Draw the quantum circuit.
+        """Draw a quantum circuit to different formats (set by output parameter):
```
Comment (1ucian0) [inline qiskit/circuit/quantumcircuit.py:1133]: The reason for the "the" is because it draws `self`.

```suggestion
        """Draw the quantum circuit. The output parameter allows to choose the drawing format:
```

File: qiskit/user_config.py [hunk 4]
```diff
@@ -29,14 +30,17 @@ class UserConfig:
...
+    transpile_optimization_level = 2
+    suppress_packaging_warnings = True
```
Comment (1ucian0) [inline qiskit/user_config.py:35]: These seems unrelated. 
```suggestion
```

File: qiskit/user_config.py [hunk 5]
```diff
@@ -60,21 +64,35 @@ def read_config_file(self):
                                           'latex_source', 'auto']:
                     raise exceptions.QiskitUserConfigError(
                         "%s is not a valid circuit drawer backend. Must be "
-                        "either 'text', 'mpl', 'latex', 'auto', or "
-                        "'latex_source'"
+                        "either 'text', 'mpl', 'latex', 'latex_source', or "
+                        "'auto'."
```
Comment (1ucian0) [inline qiskit/user_config.py:68]: what's `auto`?

File: qiskit/circuit/quantumcircuit.py [hunk 6]
```diff
@@ -1130,173 +1130,223 @@ def draw(self, output=None, scale=None, filename=None, style=None,
...
-                is no provision for passing an incomplete dict in.
-            displaycolor (dict): The color codes to use for each circuit
-                element in the form (gate_color, text_color).
-                The default values are::
+            displaycolor (dict): the color codes to use for each circuit element in
+                the form (gate_color, text_color). Colors can also be entered without
+                the text color, such as 'u1': '#FA74A6', in which case the text color
+                will always be `gatetextcolor`. The `displaycolor` dict can contain
+                any number of elements from one to the entire dict above. The default
+                values are::
```
Comment (1ucian0) [inline qiskit/circuit/quantumcircuit.py:1318]: Maybe avoid dumping the json file in the docstring?  We can refer to the file, know that there is one.

```suggestion
                any number of elements from one to the entire dict above. The default
                values are:: `default.json`
```

File: qiskit/visualization/latex.py [hunk 7]
```diff
@@ -38,8 +36,8 @@ class QCircuitImage:
     Thanks to Eric Sabo for the initial implementation for Qiskit.
     """
 
-    def __init__(self, qubits, clbits, ops, scale, style=None,
-                 plot_barriers=True, reverse_bits=False, layout=None, initial_state=False,
```
Comment (1ucian0) [inline qiskit/visualization/latex.py:42]: There were not used anyway, right? No need to deprecate-warning them?

File: qiskit/visualization/latex.py [hunk 8]
```diff
@@ -113,7 +98,6 @@ def __init__(self, qubits, clbits, ops, scale, style=None,
         # presence of "box" or "target" determines row spacing
         self.has_box = False
         self.has_target = False
-        self.reverse_bits = reverse_bits
```
Comment (1ucian0) [inline qiskit/visualization/latex.py:116]: ?

File: qiskit/visualization/matplotlib.py [hunk 9]
```diff
@@ -10,7 +10,7 @@
 # copyright notice, and modified files need to carry a notice indicating
 # that they have been altered from the originals.
 
-# pylint: disable=invalid-name,missing-docstring,inconsistent-return-statements
+# pylint: disable=invalid-name,inconsistent-return-statements
```
Comment (1ucian0) [inline qiskit/visualization/matplotlib.py:13]: nice!

File: qiskit/visualization/matplotlib.py [hunk 10]
```diff
@@ -276,16 +226,72 @@ def _registers(self, creg, qreg):
...
     @property
     def ast(self):
+        """AST getter"""
         return self._ast
 
+    def _load_style(self, style):
+        current_style = DefaultStyle().style
+        style_name = 'default'
+        config = user_config.get_config()
+        if style is False:
```
Comment (1ucian0) [inline qiskit/visualization/matplotlib.py:236]:
```suggestion
        if not style:
```

File: qiskit/visualization/matplotlib.py [hunk 11]
```diff
@@ -276,16 +226,72 @@ def _registers(self, creg, qreg):
...
+            style_name = 'bw'
+        elif isinstance(style, dict) and 'name' in style.keys():
```
Comment (1ucian0) [inline qiskit/visualization/matplotlib.py:238]:
```suggestion
        elif isinstance(style, dict) and 'name' in style:
```

File: qiskit/visualization/matplotlib.py [hunk 12]
```diff
@@ -276,16 +226,72 @@ def _registers(self, creg, qreg):
...
+            style_name = style['name']
+        elif isinstance(style, str):
+            style_name = style
+        elif config:
+            style_name = config.get('circuit_mpl_style', 'default')
```
Comment (1ucian0) [inline qiskit/visualization/matplotlib.py:243]: Maybe raising in case of `else`?
```suggestion
            style_name = config.get('circuit_mpl_style', 'default')
     else:
           raise something
```

File: qiskit/visualization/matplotlib.py [hunk 13]
```diff
@@ -276,16 +226,72 @@ def _registers(self, creg, qreg):
...
+        if style_name[-5:] == '.json':
```
Comment (1ucian0) [inline qiskit/visualization/matplotlib.py:244]: if `style_name` is shorter than 5, fails. Better 
```suggestion
        if style_name.endswith(.json'):
```

File: qiskit/visualization/matplotlib.py [hunk 14]
```diff
@@ -276,16 +226,72 @@ def _registers(self, creg, qreg):
...
+            style_name = style_name[:-5]
+
+        # Search for file in 'styles' dir, then config_path, and finally 'cwd'
+        style_path = []
+        if style_name != 'default':
+            style_name = style_name + '.json'
+            dirname = os.path.dirname(__file__)
+            style_path.append(os.path.join(dirname, 'styles', style_name))
```
Comment (1ucian0) [inline qiskit/visualization/matplotlib.py:252]: why `styles` is added?

File: qiskit/visualization/matplotlib.py [hunk 15]
```diff
@@ -276,16 +226,72 @@ def _registers(self, creg, qreg):
...
+            found_path = False
+            for path in style_path:
+                if os.path.isfile(os.path.expanduser(path)):
+                    found_path = True
+                    try:
+                        with open(os.path.expanduser(path)) as infile:
+                            json_style = json.load(infile)
+                        set_style(current_style, json_style)
+                        break
+                    except FileNotFoundError:
```
Comment (1ucian0) [inline qiskit/visualization/matplotlib.py:270]: Line 263 already checks file existance. How this exception possible?

Also, maybe `expanduser` once?

File: qiskit/visualization/matplotlib.py [hunk 16]
```diff
@@ -276,16 +226,72 @@ def _registers(self, creg, qreg):
...
+                        warn("Style JSON file '{}' not found. Will use default style.".format(
+                            style_name), UserWarning, 2)
+                    except json.JSONDecodeError as e:
+                        warn("Could not decode JSON in file '{}': {}. ".format(
+                            style_name, str(e)) + "Will use default style.", UserWarning, 2)
```
Comment (1ucian0) [inline qiskit/visualization/matplotlib.py:275]: Full file path instead of `style_name`?

File: qiskit/visualization/matplotlib.py [hunk 17]
```diff
@@ -276,16 +226,72 @@ def _registers(self, creg, qreg):
...
+                    except OSError:
+                        warn("Error loading JSON file '{}'. Will use default style.".format(
+                            style_name), UserWarning, 2)
```
Comment (1ucian0) [inline qiskit/visualization/matplotlib.py:278]: same https://github.com/Qiskit/qiskit-terra/pull/5223/files?diff=unified&w=1#r505106000

File: qiskit/visualization/matplotlib.py [hunk 18]
```diff
@@ -276,16 +226,72 @@ def _registers(self, creg, qreg):
...
+            if not found_path:
```
Comment (1ucian0) [inline qiskit/visualization/matplotlib.py:279]: Use `else:`  instead of the `found_path` flag.

File: qiskit/visualization/matplotlib.py [hunk 19]
```diff
@@ -318,8 +324,7 @@ def _get_text_width(self, text, fontsize, param=False):
                     sum_text += self._char_list['c'][f]
             return sum_text
 
-    def param_parse(self, v):
-        # create an empty list to store the parameters in
+    def _param_parse(self, v):
```
Comment (1ucian0) [inline qiskit/visualization/matplotlib.py:327]: Does the previous docstring still apply?
```suggestion
    def _param_parse(self, v):
    # create an empty list to store the parameters in
```

File: qiskit/visualization/matplotlib.py [hunk 20]
```diff
@@ -358,133 +363,137 @@ def _get_gate_ctrl_text(self, op):
...
+                gt = self._style['gt']
             else:
                 fc = color[0]
                 gt = color[1]
         # Treat special case of classical gates in iqx style by making all
         # controlled gates of x, dcx, and swap the classical gate color
-        elif self._style.name == 'iqx' and base_name in ['x', 'dcx', 'swap']:
-            color = self._style.dispcol[base_name]
+        elif self._style['name'] == 'iqx' and base_name in ['x', 'dcx', 'swap']:
+            color = self._style['dispcol'][base_name]
```
Comment (1ucian0) [inline qiskit/visualization/matplotlib.py:379]: In general, are these dictkeys mandatories?

**Review by 1ucian0 — COMMENTED** [review]
Awesome job. Do not forget the release note.

File: test/python/test_user_config.py (hunk 1 above)
Comment (enavarro51) [reply test/python/test_user_config.py]: Done in https://github.com/Qiskit/qiskit-terra/pull/5223/commits/bfb8be30672ba6d81187f57c58a4079026e0bb0e

File: qiskit/user_config.py (hunk 2 above)
Comment (enavarro51) [reply qiskit/user_config.py]: Done in https://github.com/Qiskit/qiskit-terra/pull/5223/commits/bfb8be30672ba6d81187f57c58a4079026e0bb0e

File: qiskit/circuit/quantumcircuit.py (hunk 3 above)
Comment (enavarro51) [reply qiskit/circuit/quantumcircuit.py]: Done in https://github.com/Qiskit/qiskit-terra/pull/5223/commits/bfb8be30672ba6d81187f57c58a4079026e0bb0e. Slight grammar change.

File: qiskit/user_config.py (hunk 4 above)
Comment (enavarro51) [reply qiskit/user_config.py:35]: These are all config file options. I just added them to the doc string.

File: qiskit/user_config.py (hunk 5 above)
Comment (enavarro51) [reply qiskit/user_config.py:68]: If 'auto', circuit_drawer checks _matplotlib.HAS_MATPLOTLIB. If true use 'mpl', else 'text'. Don't know if anybody uses it, but has been in code since before my time.

File: qiskit/circuit/quantumcircuit.py (hunk 6 above)
Comment (enavarro51) [reply qiskit/circuit/quantumcircuit.py]: Done in https://github.com/Qiskit/qiskit-terra/pull/5223/commits/bfb8be30672ba6d81187f57c58a4079026e0bb0e

File: qiskit/visualization/latex.py (hunk 7 above)
Comment (enavarro51) [reply qiskit/visualization/latex.py:42]: Correct. Just cleanup.

File: qiskit/visualization/latex.py (hunk 8 above)
Comment (enavarro51) [reply qiskit/visualization/latex.py:116]: Unrelated to what? reverse_bits is no longer passed, so this must be removed.

File: qiskit/visualization/matplotlib.py (hunk 9 above)
Comment (enavarro51) [reply qiskit/visualization/matplotlib.py:13]: One for the team...

File: qiskit/visualization/matplotlib.py (hunk 13 above)
Comment (enavarro51) [reply qiskit/visualization/matplotlib.py]: Done in https://github.com/Qiskit/qiskit-terra/pull/5223/commits/bfb8be30672ba6d81187f57c58a4079026e0bb0e

File: qiskit/visualization/matplotlib.py (hunk 17 above)
Comment (enavarro51) [reply qiskit/visualization/matplotlib.py]: Done in https://github.com/Qiskit/qiskit-terra/pull/5223/commits/bfb8be30672ba6d81187f57c58a4079026e0bb0e

File: qiskit/visualization/matplotlib.py (hunk 18 above)
Comment (enavarro51) [reply qiskit/visualization/matplotlib.py:277]: We iterate over a list of file paths. Not found only happens if every 'isfile' is False so 'else:' doesn't work.

File: qiskit/visualization/matplotlib.py (hunk 19 above)
Comment (enavarro51) [reply qiskit/visualization/matplotlib.py:325]: Not a docstring. Just a comment about the first line of the method. Seemed superfluous.

File: qiskit/visualization/matplotlib.py (hunk 14 above)
Comment (enavarro51) [reply qiskit/visualization/matplotlib.py:252]: We discussed making qiskit/qiskit-terra/qiskit/visualization/styles as a new directory to hold the standard styles - default, iqx, and bw.

**Comment by enavarro51** [comment]
Release note by end of day. Can I get a binder link?

File: qiskit/user_config.py (hunk 4 above)
Comment (1ucian0) [reply qiskit/user_config.py:35]: You can add them as comments too, with their default value.

File: qiskit/user_config.py (hunk 5 above)
Comment (1ucian0) [reply qiskit/user_config.py:68]: Is it documented?

File: qiskit/visualization/latex.py (hunk 8 above)
Comment (1ucian0) [reply qiskit/visualization/latex.py:116]: you are right...

File: qiskit/visualization/matplotlib.py (hunk 10 above)
Comment (1ucian0) [reply qiskit/visualization/matplotlib.py:236]: oh! I see. That's bad. It can be deprecated in a follow up.

File: qiskit/user_config.py (hunk 5 above)
Comment (enavarro51) [reply qiskit/user_config.py:68]: Not currently. May have been in the past. Functionality is in circuit_drawer, but this is only doc reference.

File: qiskit/visualization/matplotlib.py (hunk 12 above)
Comment (1ucian0) [reply qiskit/visualization/matplotlib.py:243]: sounds good to warn and continue with default if `style` is not a `str` and there is no `config`.

File: qiskit/user_config.py (hunk 5 above)
Comment (mtreinish) [reply qiskit/user_config.py:68]: It was added in https://github.com/Qiskit/qiskit-terra/pull/2331 if I remember the use correctly it was for the case where people run qiskit with the same code some where mpl works others where it didn't. So setting auto in your user config enables you to call a bare draw() and have it use mpl if available else it would fallback to text.

It wasn't documented when it was added mostly because we didn't have a good place to document it back then, and the documentation was also in a state of flux back then (it was around the time of adding multiple elements). We probably should mention of it in: https://github.com/Qiskit/qiskit/blob/master/docs/configuration.rst

File: qiskit/visualization/matplotlib.py (hunk 18 above)
Comment (1ucian0) [reply qiskit/visualization/matplotlib.py:277]: I mean something like this https://book.pythontips.com/en/latest/for_-_else.html#else-clause
You could break instead of `found_path=True`. I think, not fully sure.

File: qiskit/user_config.py (hunk 5 above)
Comment (enavarro51) [reply qiskit/user_config.py:68]: That seems like a good place to document it. We need to add `circuit_mpl_style_path` to the list as well, so I can take care of both of those.

File: qiskit/visualization/matplotlib.py (hunk 18 above)
Comment (enavarro51) [reply qiskit/visualization/matplotlib.py:277]: Oh, nice! In the 'learn something new everyday' category.

File: qiskit/visualization/matplotlib.py (hunk 12 above)
Comment (enavarro51) [reply qiskit/visualization/matplotlib.py:245]: Done in https://github.com/Qiskit/qiskit-terra/pull/5223/commits/db411c8f887c9f71a5bcd2a9455f94ab46115917

File: qiskit/visualization/matplotlib.py (hunk 18 above)
Comment (enavarro51) [reply qiskit/visualization/matplotlib.py]: Done in https://github.com/Qiskit/qiskit-terra/pull/5223/commits/db411c8f887c9f71a5bcd2a9455f94ab46115917

File: qiskit/visualization/matplotlib.py (hunk 20 above)
Comment (1ucian0) [reply qiskit/visualization/matplotlib.py:389]: could it be that, for example `self._style['dispcol']` does not exists?

**Comment by 1ucian0** [comment]
> Can I get a binder link?

You mean this? https://mybinder.org/v2/gh/enavarro51/qiskit-terra/mpl_json_loads?urlpath=apps/test/ipynb/mpl_tester.ipynb

File: qiskit/user_config.py (hunk 5 above)
Comment (1ucian0) [reply qiskit/user_config.py:68]: Maybe also here https://qiskit.org/documentation/stubs/qiskit.visualization.circuit_drawer.html?highlight=Valid%20choices%20are ? It used to be default?

File: qiskit/visualization/matplotlib.py (hunk 20 above)
Comment (enavarro51) [reply qiskit/visualization/matplotlib.py:389]: Oh, I see. `self._style = DefaultStyle().style` is loaded at the top, which includes 'dispcol' and after that calls to 'set_style' can only change the values of items, not remove them. So I think it's safe.

File: qiskit/user_config.py (hunk 5 above)
Comment (enavarro51) [reply qiskit/user_config.py:68]: Actually, I think this is only for the user_config based on @mtreinish first paragraph comment above. A user wouldn't use it if calling QuantumCircuit.draw or circuit_drawer.

**Review by 1ucian0 — APPROVED** [review]
Thanks! Great work!

**Comment by enavarro51** [comment]
Thanks. It was fun.
